connectors/idp-extraction/src/main/java/io/camunda/connector/idp/extraction/caller/PollingTextractCaller.java

```diff
@@ -115,35 +181,75 @@ private List<Block> processDocument(
     return allBlocks;
   }
 
-  private TextractTask prepareTextractTextDetectionTask(
+  private TextractAnalysisTask prepareTextractAnalysisTask(
+      String jobId, TextractClient textractClient) {
+    GetDocumentAnalysisRequest documentAnalysisRequest =
+        GetDocumentAnalysisRequest.builder().jobId(jobId).maxResults(MAX_RESULT).build();
+
+    return new TextractAnalysisTask(documentAnalysisRequest, textractClient);
+  }
+
+  private GetDocumentAnalysisResponse executeAnalysisTask(
+      TextractAnalysisTask task, long delay, ScheduledExecutorService executorService)
+      throws Exception {
+    ScheduledFuture<GetDocumentAnalysisResponse> nextDocumentResultFuture =
+        executorService.schedule(task, delay, SECONDS);
+    return nextDocumentResultFuture.get();
+  }
+
+  private TextractTextDetectionTask prepareTextractTextDetectionTask(
       String jobId, TextractClient textractClient) {
     GetDocumentTextDetectionRequest documentTextDetectionRequest =
         GetDocumentTextDetectionRequest.builder().jobId(jobId).maxResults(MAX_RESULT).build();
 
-    return new TextractTask(documentTextDetectionRequest, textractClient);
+    return new TextractTextDetectionTask(documentTextDetectionRequest, textractClient);
   }
 
   private GetDocumentTextDetectionResponse executeTextDetectionTask(
-      TextractTask task, long delay, ScheduledExecutorService executorService) throws Exception {
+      TextractTextDetectionTask task, long delay, ScheduledExecutorService executorService)
+      throws Exception {
     ScheduledFuture<GetDocumentTextDetectionResponse> nextDocumentResultFuture =
         executorService.schedule(task, delay, SECONDS);
     return nextDocumentResultFuture.get();
   }
 
-  private StructuredExtractionResponse extractKeyValuePairsWithConfidence(List<Block> blocks) {
+  private StructuredExtractionResponse extractDataFromDocument(List<Block> blocks) {
     Map<String, String> keyValuePairs = new HashMap<>();
     Map<String, Float> confidenceScores = new HashMap<>();
     Map<String, Block> blockMap =
         blocks.stream().collect(Collectors.toMap(Block::id, block -> block));
+    Map<String, Integer> keyOccurrences = new HashMap<>();
+
+    extractKeyValueSet(blocks, blockMap, keyValuePairs, keyOccurrences, confidenceScores);
 
+    return new StructuredExtractionResponse(keyValuePairs, confidenceScores);
+  }
+
+  private void extractKeyValueSet(
+      List<Block> blocks,
+      Map<String, Block> blockMap,
+      Map<String, String> keyValuePairs,
+      Map<String, Integer> keyOccurrences,
+      Map<String, Float> confidenceScores) {
```

Comment:
🔧 It's pretty hard to follow that the maps which are passed into this method are mutable and being mutated within this method. Is there really a point in extracting this logic into its own method opposed to keeping it in `extractDataFromDocument` where the variables are initialized? Why does `extractDataFromDocument` need to know about `keyOccurrences` if they are only used inside `extractKeyValueSet`?

I would suggest to either:

- keep the logic in the same method
- or only pass the needed data (the blocks) to the `extractKeyValueSet` method and to return relevant results (`keyValuePairs`, `confidenceScores`) in a some data structure/record